### PR TITLE
tee_svc_hash_final(): reset state before returning to caller

### PR DESCRIPTION
**Pull request by jforissier**
As per GlobalPlatform Internal Core API Specification v1.1, section
6.3.2 TEE_DigestDoFinal: "Afterwards the Message Digest operation is
reset to initial state and can be reused".

Signed-off-by: Jerome Forissier <jerome.forissier@linaro.org>

### CODE REVIEW

**Comment by pascal-brand-st-dev** [comment]
* Indeed, this is not a GP 1.1 specific requirement. It was already there in v1.0. So the commit message should be updated.
* Cedric is adding such changes in lib/libutee/tee_api_operation.c. I think it is cleaner the way Cédric is adding things as all "GP" related stuff is in libutee.
  * we better know all the changes. Fro ex., your change does not apply on `TEE_DigestDoFinal`, but to `TEE_MACComputeFinal` too
  * but I agree that this is not optimal in terms of ioctl, and that we also have to add this change for `TEE_MACComputeFinal`. But I think we better know all the changes

**Comment by jforissier** [comment]
OK, I will update the commit message and move the code to libutee.

**Comment by jforissier** [comment]
Updated.

**Comment by pascal-brand-st-dev** [comment]
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

**Comment by jforissier** [comment]
No I have a doubt about `TEE_MACComputeFinal()`. In fact I think we should not reset the state.

GP states that a panic reason is "The operation has not been started yet with TEE_MACInit or has already been finalized.", and "The TEE_MACComputeFinal function finalizes the MAC operation...". So it seems you must call `TEE_MACInit()` again before you can call `TEE_MACComputeFinal()` again.
Same goes for the other "final" functions:
- `TEE_MACCompareFinal()`, TEE_CipherDoFinal()` have the same wording as above
- `TEE_AEEncryptFinal()` and `TEE_AEDecryptFinal()` just have "the operation has not started yet" as a panic reason, but I suspect it is the same intent as they seem in every way similar to the above
- Only `TEE_DigestDoFinal()` would need to be fixed because it is the only one that has no "init" function.
In other words, all operations can be reused but it is always necessary to call `TEE_Initxxx()` again except for digest which has no such init function.

So, I suggest I change only `TEE_DigestDoFinal()` in libutee, i.e. drop the changes in `TEE_MACComputeFinal()`. Does it sound good?

**Comment by pascal-brand-st-dev** [comment]
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

Ok with the change only in `TEE_DigestDoFinal`

**Comment by jforissier** [comment]
Rebased and review tag applied.
